odoo/http.py

```diff
@@ -986,18 +986,18 @@ def authenticate(self, dbname, login=None, password=None):
         }
 
         registry = Registry(dbname)
-        pre_uid = registry['res.users'].authenticate(dbname, login, password, wsgienv)
+        pre_uid = registry['res.users'].authenticate(dbname, credential, wsgienv)
 
         self.uid = None
-        self.pre_login = login
+        self.pre_login = credential['login']
         self.pre_uid = pre_uid
 
         with registry.cursor() as cr:
             env = odoo.api.Environment(cr, pre_uid, {})
 
             # if 2FA is disabled we finalize immediately
             user = env['res.users'].browse(pre_uid)
-            if not user._mfa_url():
+            if credential.get('skip_totp') or not user._mfa_url():
```

Comment:
Doesn't that mean that if I use as credential dict:
`{'type': 'password', 'login': 'admin', 'password': 'admin', 'skip_totp': True}`

I can escape the 2FA authentication when using login/password ?

You could avoid this by forcing to `credential['skip_totp'] = True` somewhere, and then make it `False` when going through the pass key authentication. But, still, this is highly fragile/tricky, it really requires the developer to be really careful when he does an overrideide of `authenticate`/`_check_credentials`. Especially with the mutable trait of the dict.

Now that I see this I am even wondering if we shouldn't protect the `credential` dict with a `frozendict` in `authenticate`, to protect from changes in the underlying calls to `_login`  and `_check_credential` to cover one more attack vector.